File: 주은/W2/전력망을_둘로_나누기.cpp
#include <string>
#include <vector>
#include <climits>
#include <cstdlib>

using namespace std;
int check[110];
int line[110][110];
int cnt = 0, cnt_a, cnt_b;
int minn = INT_MAX;

void dfs(int n, int x, int y)
{
    check[x] = 1;
    cnt++;
    
    for (int i = 1; i <= n; i++) {
        if (line[x][i] == 1 && check[i] == 0 && i != y) {
            dfs(n, i, y);
        }
    }
}

int solution(int n, vector<vector<int>> wires) {
    int answer = -1;
    int a, b;
    for (int i = 0; i < wires.size(); i++) {
        line[wires[i][0]][wires[i][1]] = 1;
        line[wires[i][1]][wires[i][0]] = 1;
    }
    
    for (int i = 0; i < wires.size(); i++) {
        a = wires[i][0];
        b = wires[i][1];
        for (int j = 1; j <= n; j++) check[j] = 0;
        cnt = 0;
        dfs(n, a, b);
        cnt_a = cnt;
        cnt = 0;
        dfs(n, b, a);
        cnt_b = cnt;
        if (minn > abs(cnt_a-cnt_b)) minn = abs(cnt_a-cnt_b);
    }
    
    answer = minn;
    return answer;
}

File: 서연/W2/전력망을 둘로 나누기.py
def solution(n, wires):
    answer = n
    
    for i in range(len(wires)):
        sub_wires = wires[:i] + wires[i+1:]
        
        cluster = {wires[i][0]}
        
        for _ in range(len(sub_wires)):
            for v1, v2 in sub_wires:
                if v1 in cluster: cluster.add(v2)
                if v2 in cluster: cluster.add(v1)
        
        count = len(cluster)
        answer = min(answer, abs(count - (n - count)))
        
    return answer
